### PR TITLE
Switch to using cluster-hosted Ironic for worker deployments

### PR DESCRIPTION
**Pull request by hardys**
Testing with some additional patches to prove #635 in CI cc @imain

### CODE REVIEW

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/864/

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/866/

File: 08_deploy_bmo.sh [hunk 1]
```diff
@@ -12,10 +12,36 @@ export BMOPATH="$GOPATH/src/github.com/metal3-io/baremetal-operator"
 # Make a local copy of the baremetal-operator code to make changes
 cp -r $BMOPATH/deploy ocp/.
 sed -i 's/namespace: .*/namespace: openshift-machine-api/g' ocp/deploy/role_binding.yaml
+cp $SCRIPTDIR/operator_ironic.yaml ocp/deploy
+cp $SCRIPTDIR/ironic_bmo_configmap.yaml ocp/deploy
+sed -i "s#__RHCOS_IMAGE_URL__#${RHCOS_IMAGE_URL}#" ocp/deploy/ironic_bmo_configmap.yaml
+
+# Kill the dnsmasq container on the host since it is performing DHCP and doesn't
+# allow our pod in openshift to take over.
+for name in dnsmasq ironic-inspector ; do
```
Comment (russellb) [inline 08_deploy_bmo.sh]: What about the rest of the ironic containers?

File: 08_deploy_bmo.sh [hunk 2]
```diff
@@ -12,10 +12,36 @@ export BMOPATH="$GOPATH/src/github.com/metal3-io/baremetal-operator"
...
+oc --config ocp/auth/kubeconfig create secret generic mariadb-password --from-literal password=password --namespace=openshift-machine-api
+
+oc --config ocp/auth/kubeconfig adm --as system:admin policy add-scc-to-user privileged system:serviceaccount:openshift-machine-api:baremetal-operator
+oc --config ocp/auth/kubeconfig apply -f ocp/deploy/operator_ironic.yaml -n openshift-machine-api
+
+# Sadly I don't see a way to get this from the json..
+POD_NAME=$(oc --config ocp/auth/kubeconfig get pods -n openshift-machine-api | grep metal3-baremetal-operator | cut -f 1 -d ' ')
+
+# Make sure our pod is running.
+echo "Waiting for baremetal-operator pod to become ready"
```
Comment (russellb) [inline 08_deploy_bmo.sh]: Why do we need to wait?  It shouldn't be required to block here

File: 08_deploy_bmo.sh [hunk 3]
```diff
@@ -12,10 +12,36 @@ export BMOPATH="$GOPATH/src/github.com/metal3-io/baremetal-operator"
...
+
+oc --config ocp/auth/kubeconfig apply -f ocp/deploy/ironic_bmo_configmap.yaml --namespace=openshift-machine-api
+# I'm leaving this as is for debugging but we could easily generate a random password here.
+oc --config ocp/auth/kubeconfig delete secret mariadb-password --namespace=openshift-machine-api || true
+oc --config ocp/auth/kubeconfig create secret generic mariadb-password --from-literal password=password --namespace=openshift-machine-api
+
+oc --config ocp/auth/kubeconfig adm --as system:admin policy add-scc-to-user privileged system:serviceaccount:openshift-machine-api:baremetal-operator
+oc --config ocp/auth/kubeconfig apply -f ocp/deploy/operator_ironic.yaml -n openshift-machine-api
+
+# Sadly I don't see a way to get this from the json..
```
Comment (russellb) [inline 08_deploy_bmo.sh]: `oc get pod -l name=metal3-baremetal-operator -n openshift-machine-api -o jsonpath="{.items[0].metadata.name}"`

File: 08_deploy_bmo.sh (hunk 1 above)
Comment (imain) [reply 08_deploy_bmo.sh]: If we kill all the containers we lose the ironic database and cleanup will not work.  eg 'make clean' in dev-scripts will fail to remove the ironic managed masters.

File: 08_deploy_bmo.sh (hunk 2 above)
Comment (imain) [reply 08_deploy_bmo.sh]: Yeah it's probably optional.  I haven't tested the whole thing without the wait yet though.  I'll give it a go.  On the other hand we could use this to catch errors early.

Comment (russellb) [reply 08_deploy_bmo.sh]: Fair enough.

File: 08_deploy_bmo.sh (hunk 1 above)
Comment (russellb) [reply 08_deploy_bmo.sh]: OK, expanding the comment would help.  I imagine all of this is going away shortly anyway once the bootstrap VM ironic work lands.

File: 08_deploy_bmo.sh (hunk 2 above)
Comment (hardys) [reply 08_deploy_bmo.sh]: FWIW in my testing this was helpful when one of the containers went into CrashLoopBackoff, it meant I could start investigating when it became clear the pod was wedged and not starting correctly.

I don't have a strong opinion but for development having some verbose monitoring of the pod startup is probably no bad thing?

File: 08_deploy_bmo.sh (hunk 1 above)
Comment (hardys) [reply 08_deploy_bmo.sh]: Yeah that's correct, but then we'll also need to modify make clean to not rely on the terraform cleanup (until we figure out how to reimplement destroy)

**Comment by hardys** [comment]
@imain how do you want to proceed with this PR vs https://github.com/openshift-metal3/dev-scripts/pull/635 ?

I pushed this mainly to prove things in CI, but if you're happy with the approach of breaking the dependency on https://github.com/metal3-io/baremetal-operator/pull/212 we could go ahead with this one? wdyt?

File: 08_deploy_bmo.sh (hunk 2 above)
Comment (imain) [reply 08_deploy_bmo.sh]: yeah I'm 50/50 on it.  I do kinda like how it can catch an issue with the pod, but for CI it's probably not useful.

**Comment by russellb** [comment]
It’s fine. It can stay for the dev scripts. A worker couldn’t deploy until
this is done anyway.

Sent from mobile device

On Jul 10, 2019, at 7:22 PM, Ian Main <notifications@github.com> wrote:

*@imain* commented on this pull request.
------------------------------

In 08_deploy_bmo.sh
<https://github.com/openshift-metal3/dev-scripts/pull/659#discussion_r302307008>
:

>
 # Start deploying on the new cluster
 oc --config ocp/auth/kubeconfig apply -f
ocp/deploy/service_account.yaml --namespace=openshift-machine-api
 oc --config ocp/auth/kubeconfig apply -f ocp/deploy/role.yaml
--namespace=openshift-machine-api
 oc --config ocp/auth/kubeconfig apply -f ocp/deploy/role_binding.yaml
 oc --config ocp/auth/kubeconfig apply -f
ocp/deploy/crds/metal3_v1alpha1_baremetalhost_crd.yaml
-oc --config ocp/auth/kubeconfig apply -f ocp/deploy/operator.yaml
--namespace=openshift-machine-api
+
+oc --config ocp/auth/kubeconfig apply -f
ocp/deploy/ironic_bmo_configmap.yaml --namespace=openshift-machine-api
+# I'm leaving this as is for debugging but we could easily generate a
random password here.
+oc --config ocp/auth/kubeconfig delete secret mariadb-password
--namespace=openshift-machine-api || true
+oc --config ocp/auth/kubeconfig create secret generic
mariadb-password --from-literal password=password
--namespace=openshift-machine-api
+
+oc --config ocp/auth/kubeconfig adm --as system:admin policy
add-scc-to-user privileged
system:serviceaccount:openshift-machine-api:baremetal-operator
+oc --config ocp/auth/kubeconfig apply -f
ocp/deploy/operator_ironic.yaml -n openshift-machine-api
+
+# Sadly I don't see a way to get this from the json..
+POD_NAME=$(oc --config ocp/auth/kubeconfig get pods -n
openshift-machine-api | grep metal3-baremetal-operator | cut -f 1 -d '
')
+
+# Make sure our pod is running.
+echo "Waiting for baremetal-operator pod to become ready"

yeah I'm 50/50 on it. I do kinda like how it can catch an issue with the
pod, but for CI it's probably not useful.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub
<https://github.com/openshift-metal3/dev-scripts/pull/659?email_source=notifications&email_token=AACLQCRRRKMX25AJ4JMBZCLP6ZVJVA5CNFSM4H7FH3HKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6CWCUY#discussion_r302307008>,
or mute the thread

**Comment by hardys** [comment]
@imain is updating #635 so I'll abandon this one

**Pull request closed without merging**